File: internal/definition/bundle.go
package definition

import "fmt"

// ValidateBundle validates every definition in a bundle, rejects duplicate
// slugs, and checks that each form's workflow resolves to a workflow in the
// bundle or in existingWorkflowSlugs. Problem paths are prefixed with
// "forms[i]" / "workflows[i]".
func ValidateBundle(forms []Form, workflows []Workflow, existingWorkflowSlugs []string) error {
	var ps problems
	known := make(map[string]bool, len(existingWorkflowSlugs)+len(workflows))
	for _, s := range existingWorkflowSlugs {
		known[s] = true
	}

	seenW := make(map[string]bool, len(workflows))
	for i, w := range workflows {
		p := fmt.Sprintf("workflows[%d]", i)
		ps.merge(p, ValidateWorkflow(w))
		if seenW[w.Slug] {
			ps.add(p+".slug", fmt.Sprintf("duplicate workflow slug %q", w.Slug))
		}
		seenW[w.Slug] = true
		known[w.Slug] = true
	}

	seenF := make(map[string]bool, len(forms))
	for i, f := range forms {
		p := fmt.Sprintf("forms[%d]", i)
		ps.merge(p, ValidateForm(f))
		if seenF[f.Slug] {
			ps.add(p+".slug", fmt.Sprintf("duplicate form slug %q", f.Slug))
		}
		seenF[f.Slug] = true
		if f.Workflow != "" && !known[f.Workflow] {
			ps.add(p+".workflow", fmt.Sprintf("unknown workflow %q", f.Workflow))
		}
	}
	return ps.err()
}

File: src/openforms/definition/bundle.py
"""Bundle validation: every definition, duplicate slugs and workflow references."""

from __future__ import annotations

from collections.abc import Iterable

from .problems import Problems, ValidationError
from .types import Form, Workflow
from .validate_form import validate_form
from .validate_workflow import validate_workflow


def _err(fn, arg) -> ValidationError | None:
    try:
        fn(arg)
    except ValidationError as e:
        return e
    return None


def validate_bundle(forms: list[Form], workflows: list[Workflow], existing_workflow_slugs: Iterable[str] = ()) -> None:
    """Problem paths are prefixed ``forms[i]`` / ``workflows[i]``."""
    ps = Problems()
    known = set(existing_workflow_slugs)
    seen_w: set[str] = set()
    for i, w in enumerate(workflows):
        p = f"workflows[{i}]"
        ps.merge(p, _err(validate_workflow, w))
        if w.slug in seen_w:
            ps.add(p + ".slug", f'duplicate workflow slug "{w.slug}"')
        seen_w.add(w.slug)
        known.add(w.slug)
    seen_f: set[str] = set()
    for i, f in enumerate(forms):
        p = f"forms[{i}]"
        ps.merge(p, _err(validate_form, f))
        if f.slug in seen_f:
            ps.add(p + ".slug", f'duplicate form slug "{f.slug}"')
        seen_f.add(f.slug)
        if f.workflow and f.workflow not in known:
            ps.add(p + ".workflow", f'unknown workflow "{f.workflow}"')
    ps.raise_if_any()
